src/shared/openssl-util.c

```diff
@@ -137,4 +223,159 @@ int string_hashsum(
 
 }
 #  endif
+
+/* We use derived shared secret as volume key for EC keys.
+ * The following is the procedure.
+ * First we generate a pair of EC key of the same curve group.
+ * Then we use newly generated private key to derive shared secret
+ * with the public key in the token.
+ * Finally, we simplely forget the private key, store the public key
+ * to pkcs11-key field of LUKS token and use the dervied shared secret
+ * as the volume key (we don't apply any KDF here since LUKS will apply
+ * PBKDF on the key).
+ */
+static int pkey_generate_volume_key_ec(
+                EVP_PKEY *pkey,
+                void **ret_decrypted_key,
+                size_t *ret_decrypted_key_size,
+                void **ret_savedata,
+                size_t *ret_savedata_size) {
+
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *pkey_new = NULL;
+        int nid = NID_undef;
+        int r;
+
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+        _cleanup_free_ char *curve_name = NULL;
+        size_t len = 0;
+        if (EVP_PKEY_get_group_name(pkey, NULL, 0, &len) != 1 || len == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to determine PKEY group name length");
+        len += 1;
+        curve_name = malloc(len);
+        if (!curve_name)
+                return log_oom();
+
+        if (!EVP_PKEY_get_group_name(pkey, curve_name, len, &len))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get PKEY group name");
+        nid = OBJ_sn2nid(curve_name);
+#else
+        EC_KEY *ec_key = EVP_PKEY_get0_EC_KEY(pkey);
+        if (!ec_key)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "PKEY doesn't have EC_KEY associated");
+
+        if (EC_KEY_check_key(ec_key) != 1)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "EC_KEY associated with PKEY is not valid");
+
+        nid = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec_key));
+#endif
+
+        r = pkey_generate_ec_key(nid, &pkey_new);
+        if (r < 0)
+                return log_error_errno(r, "Failed to generate ec key: %m");
+
+        r = pkey_ecdh_derive_shared_secret(pkey_new, pkey, NULL, ret_decrypted_key_size);
+        if (r < 0 || *ret_decrypted_key_size == 0)
+                return log_error_errno(r, "Failed to determine derived shared secret size: %m");
```

Comment:
if `pkey_ecdh_derive_shared_secret()` succeeds, but `*ret_decrypted_key_size` is zero, then you'll hit this error path, with r == 0, and propagate that as error, but zero is no error. This should be two error paths, one `r < 0` and one `*ret_decrypted_key_size == 0`

Reply:
moreover, we have this rule in systemd that functions should not touch their return parameters on failure, and should initialize them fully on success. It's not OK to modify return values on failure. Here, `*ret_decrypted_key_size` might be initialized, but then later the function might still fail, conflicting with the rule.

Hence, please make pkey_ecdh_derive_shared_secret() write its result into a local temporary variable first, and then when nothing can fail anymore, copy it over to the actual return parameter.